File: examples/python/example_agent_huggingface.py
#!/usr/bin/env python3
"""Option A example — GoodMem collection wired into a Semantic Kernel agent (HuggingFace).

The agent has a memory search tool backed by GoodMem.  When the user asks
a question, the LLM decides whether to call the tool to look up relevant
memories before composing its answer.

This is the HuggingFace variant of example_agent.py.  Compare the two files
side-by-side: everything from the GoodMem import onward is identical — only
the import and service= line differ.  That is intentional: GoodMem is
provider-agnostic; swapping the LLM requires changing exactly one line.

NOTE — why this uses OpenAIChatCompletion rather than HuggingFaceTextCompletion:
    SK's built-in HuggingFace connector (HuggingFaceTextCompletion) runs models
    locally via the transformers pipeline and implements TextCompletionClientBase,
    not ChatCompletionClientBase.  ChatCompletionAgent requires a chat interface,
    so the local connector cannot be used here directly.

    The HuggingFace Inference API (https://api-inference.huggingface.co/v1/) is
    OpenAI-compatible, so we point OpenAIChatCompletion at it instead.  The model
    runs on HuggingFace infrastructure; no local GPU is required.  Any model on
    the Hub that supports the Messages API works.

Requirements (in addition to goodmem-sk):
    pip install semantic-kernel openai

Environment variables:
    GOODMEM_BASE_URL    — GoodMem server URL  (default: http://localhost:8080)
    GOODMEM_VERIFY_SSL  — Set to 'false' for self-signed certs
    GOODMEM_API_KEY     — GoodMem API key
    HF_TOKEN            — HuggingFace API token (https://huggingface.co/settings/tokens)
    HF_MODEL            — HuggingFace model ID (default: meta-llama/Llama-3.3-70B-Instruct)
                          Must support the Messages API (chat/instruction-tuned models).

Usage:
    GOODMEM_BASE_URL=https://localhost:8080 \
    GOODMEM_VERIFY_SSL=false \
    GOODMEM_API_KEY=your-key \
    HF_TOKEN=hf_... \
    HF_MODEL=meta-llama/Llama-3.3-70B-Instruct \
    python example_agent_huggingface.py
"""

import asyncio
import os
from dataclasses import dataclass
from typing import Annotated

from openai import AsyncOpenAI

from semantic_kernel.agents import AgentThread, ChatCompletionAgent
from semantic_kernel.connectors.ai import FunctionChoiceBehavior
from semantic_kernel.connectors.ai.open_ai import OpenAIChatCompletion
from semantic_kernel.data.vector import VectorStoreField, vectorstoremodel
from semantic_kernel.functions import KernelParameterMetadata, KernelPlugin

from goodmem_semantic_kernel import GoodMemCollection


_HF_INFERENCE_BASE_URL = "https://api-inference.huggingface.co/v1/"
_DEFAULT_HF_MODEL = "meta-llama/Llama-3.3-70B-Instruct"


# ---------------------------------------------------------------------------
# Data model
# ---------------------------------------------------------------------------

@vectorstoremodel
@dataclass
class Memory:
    id: Annotated[str | None, VectorStoreField("key")] = None
    content: Annotated[str, VectorStoreField("data", type="str")] = ""
    topic: Annotated[str | None, VectorStoreField("data")] = None


# ---------------------------------------------------------------------------
# Seed data
# ---------------------------------------------------------------------------

SEED_MEMORIES = [
    Memory(content="The Eiffel Tower is located in Paris, France.", topic="geography"),
    Memory(content="Python was created by Guido van Rossum and first released in 1991.", topic="technology"),
    Memory(content="The speed of light is approximately 299,792 km/s.", topic="science"),
    Memory(content="Shakespeare wrote Hamlet, Macbeth, and Romeo and Juliet.", topic="literature"),
    Memory(content="The Pacific Ocean is the largest ocean on Earth.", topic="geography"),
    Memory(content="Semantic Kernel is a Microsoft SDK for building AI agents.", topic="technology"),
]


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

async def main() -> None:
    for var in ("GOODMEM_API_KEY", "HF_TOKEN"):
        if not os.environ.get(var):
            raise SystemExit(f"Set {var} before running this script.")

    hf_model = os.environ.get("HF_MODEL", _DEFAULT_HF_MODEL)

    # Point OpenAIChatCompletion at HuggingFace's OpenAI-compatible Inference API.
    # The AsyncOpenAI client is passed in so SK does not try to read OPENAI_API_KEY.
    hf_client = AsyncOpenAI(
        base_url=_HF_INFERENCE_BASE_URL,
        api_key=os.environ["HF_TOKEN"],
    )
    service = OpenAIChatCompletion(ai_model_id=hf_model, async_client=hf_client)

    async with GoodMemCollection(record_type=Memory, collection_name="agent-memory") as collection:
        # 1. Fresh space with seed data
        await collection.ensure_collection_deleted()
        await collection.ensure_collection_exists()
        await collection.upsert(SEED_MEMORIES)
        print(f"Seeded {len(SEED_MEMORIES)} memories into 'agent-memory'.")

        print("Waiting for embeddings...")
        await asyncio.sleep(3)

        # 2. Turn the collection into a kernel plugin the agent can call.
        #    create_search_function() builds a callable that the LLM sees as a
        #    tool — it calls collection.search(query, top=N) when invoked.
        memory_plugin = KernelPlugin(
            name="memory",
            description="Long-term memory store. Search it to recall facts.",
            functions=[
                collection.create_search_function(
                    function_name="recall",
                    description=(
                        "Search long-term memory for facts relevant to a query. "
                        "Call this before answering questions about facts, geography, "
                        "science, technology, or literature."
                    ),
                    parameters=[
                        KernelParameterMetadata(
                            name="query",
                            description="What to search for in memory.",
                            type="str",
                            is_required=True,
                            type_object=str,
                        ),
                        KernelParameterMetadata(
                            name="top",
                            description="Number of memories to retrieve (default 3).",
                            type="int",
                            default_value=3,
                            type_object=int,
                        ),
                    ],
                    # Format each result as a single line shown to the LLM
                    string_mapper=lambda r: r.record.content,
                ),
            ],
        )

        # 3. Build the agent
        agent = ChatCompletionAgent(
            name="MemoryAgent",
            service=service,
            instructions=(
                "You are a helpful assistant with access to a long-term memory store. "
                "Always search memory before answering factual questions. "
                "Cite what you found in memory when it is relevant."
            ),
            function_choice_behavior=FunctionChoiceBehavior.Auto(),
            plugins=[memory_plugin],
        )

        # 4. Interactive chat loop
        print(f"\nMemory agent ready (model: {hf_model}). Type 'exit' to quit.\n")
        thread: AgentThread | None = None
        while True:
            try:
                user_input = input("You: ").strip()
            except (KeyboardInterrupt, EOFError):
                print("\nBye!")
                break

            if not user_input or user_input.lower() == "exit":
                break

            result = await agent.get_response(messages=user_input, thread=thread)
            thread = result.thread
            print(f"Agent: {result.content}\n")

        # 5. Uncomment to clean up:
        # await collection.ensure_collection_deleted()
        # print("Memory space deleted.")


if __name__ == "__main__":
    asyncio.run(main())

File: samples/java/ExampleAgentHuggingFace/src/main/java/ai/goodmem/samples/ExampleAgentHuggingFace.java
package ai.goodmem.samples;

// ExampleAgentHuggingFace — GoodMem collection wired into a Semantic Kernel agent (Hugging Face).
//
// The agent has a memory search tool backed by GoodMem. When the user asks a
// question, the LLM decides whether to call the tool to look up relevant
// memories before composing its answer.
//
// Note: SK's built-in HuggingFace connector does not support chat completion
// with function calling. This sample uses HuggingFace's OpenAI-compatible
// Inference API endpoint instead, via the SK OpenAI connector with a URL-rewrite
// policy. Not all HuggingFace models support tool calling — instruction-tuned
// models work best (e.g. Llama-3.1-8B-Instruct).
//
// Required environment variables:
//   GOODMEM_BASE_URL   — GoodMem server URL  (default: http://localhost:8080)
//   GOODMEM_VERIFY_SSL — Set to 'false' for self-signed certs
//   GOODMEM_API_KEY    — GoodMem API key
//   HF_TOKEN           — Hugging Face API token (https://huggingface.co/settings/tokens)
//
// Optional environment variables:
//   HF_MODEL           — Hugging Face model ID (default: meta-llama/Llama-3.1-8B-Instruct)
//
// Run:
//   cd samples/java/ExampleAgentHuggingFace
//   mvn install -f ../../../java/pom.xml -DskipTests
//   mvn compile exec:java

import ai.goodmem.semantickernel.GoodMemCollection;
import ai.goodmem.semantickernel.GoodMemData;
import ai.goodmem.semantickernel.GoodMemKey;
import ai.goodmem.semantickernel.GoodMemPlugin;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;
import reactor.netty.http.HttpResources;
import com.microsoft.semantickernel.Kernel;
import com.microsoft.semantickernel.aiservices.openai.chatcompletion.OpenAIChatCompletion;
import com.microsoft.semantickernel.orchestration.InvocationContext;
import com.microsoft.semantickernel.orchestration.ToolCallBehavior;
import com.microsoft.semantickernel.plugin.KernelPluginFactory;
import com.microsoft.semantickernel.services.chatcompletion.ChatCompletionService;
import com.microsoft.semantickernel.services.chatcompletion.ChatHistory;
import com.azure.ai.openai.OpenAIAsyncClient;
import com.azure.ai.openai.OpenAIClientBuilder;
import com.azure.core.credential.KeyCredential;
import com.azure.core.http.HttpPipelineCallContext;
import com.azure.core.http.HttpPipelineNextPolicy;
import com.azure.core.http.HttpResponse;
import com.azure.core.http.policy.HttpPipelinePolicy;

import java.util.List;
import java.util.Scanner;

public class ExampleAgentHuggingFace {

    // HuggingFace Inference API uses an OpenAI-compatible endpoint.
    // Azure SDK forces Azure-mode paths when .endpoint() is set, so we stay in
    // non-Azure mode and rewrite api.openai.com → router.huggingface.co at the HTTP layer.
    private static final String OPENAI_BASE = "https://api.openai.com/v1";
    private static final String HF_BASE     = "https://router.huggingface.co/v1";
    private static final String DEFAULT_HF_MODEL = "meta-llama/Llama-3.1-8B-Instruct";

    // ── Data model ────────────────────────────────────────────────────────────
    public static class Memory {
        @GoodMemKey
        public String id;

        @GoodMemData                   // becomes originalContent
        public String content;

        @GoodMemData("source")         // stored as metadata["source"]
        public String source;
    }

    public static void main(String[] args) throws InterruptedException {
        for (var v : List.of("GOODMEM_API_KEY", "GOODMEM_BASE_URL", "HF_TOKEN")) {
            if (System.getenv(v) == null || System.getenv(v).isBlank())
                throw new IllegalStateException("Set " + v + " before running this example.");
        }

        var model = System.getenv("HF_MODEL");
        if (model == null || model.isBlank()) model = DEFAULT_HF_MODEL;

        // ── 1. Set up GoodMem collection with seed data ───────────────────────
        var collection = GoodMemCollection.of("agent-memory", Memory.class);

        collection.ensureCollectionDeleted().block();
        collection.ensureCollectionExists().block();
        collection.upsertAll(List.of(
                memory("The Pacific Ocean is the largest ocean on Earth.", "geography"),
                memory("Python was created by Guido van Rossum and first released in 1991.", "technology"),
                memory("The speed of light is approximately 299,792 km/s.", "science"),
                memory("Shakespeare wrote Hamlet, Macbeth, and Romeo and Juliet.", "literature"),
                memory("Semantic Kernel is a Microsoft SDK for building AI agents.", "technology")
        )).blockLast();
        System.out.println("Seeded 5 memories into the 'agent-memory' GoodMem space.");

        System.out.println("Waiting for embeddings...");
        Thread.sleep(3_000);

        // ── 2. Build the Semantic Kernel using HF's OpenAI-compatible Inference API ──
        OpenAIAsyncClient hfClient = new OpenAIClientBuilder()
                .credential(new KeyCredential(System.getenv("HF_TOKEN")))
                .addPolicy(new HuggingFaceUrlRewritePolicy())
                .buildAsyncClient();

        ChatCompletionService chatService = OpenAIChatCompletion.builder()
                .withOpenAIAsyncClient(hfClient)
                .withModelId(model)
                .build();

        // ── 3. Register GoodMemPlugin so the LLM can call memory.recall ───────
        var memoryPlugin = new GoodMemPlugin<>(collection, Memory.class, (content, proto) -> {
            Memory m = new Memory();
            m.content = content;
            m.source = "agent";
            return m;
        });

        var kernel = Kernel.builder()
                .withAIService(ChatCompletionService.class, chatService)
                .withPlugin(KernelPluginFactory.createFromObject(memoryPlugin, "memory"))
                .build();

        // ── 4. Interactive chat loop ───────────────────────────────────────────
        System.out.println("\nMemory agent ready (model: " + model + "). Type 'exit' to quit.\n");

        var history = new ChatHistory(
                "You are a helpful assistant with access to a long-term memory store. "
                + "Always search memory before answering factual questions. "
                + "Cite what you found in memory when it is relevant.");

        var invocationContext = InvocationContext.builder()
                .withToolCallBehavior(ToolCallBehavior.allowAllKernelFunctions(true))
                .build();

        var scanner = new Scanner(System.in);
        while (true) {
            System.out.print("You: ");
            var input = scanner.nextLine().trim();
            if (input.isEmpty() || "exit".equalsIgnoreCase(input)) break;

            history.addUserMessage(input);

            var response = chatService
                    .getChatMessageContentsAsync(history, kernel, invocationContext)
                    .block();

            if (response != null && !response.isEmpty()) {
                var reply = response.get(response.size() - 1).getContent();
                System.out.println("Agent: " + reply);
                history.addAssistantMessage(reply);
            }
        }

        collection.close();
        HttpResources.disposeLoopsAndConnections();
        Schedulers.shutdownNow();
    }

    private static Memory memory(String content, String source) {
        Memory m = new Memory();
        m.content = content;
        m.source = source;
        return m;
    }

    /** Rewrites api.openai.com requests to the HuggingFace Inference API at the HTTP layer. */
    static class HuggingFaceUrlRewritePolicy implements HttpPipelinePolicy {
        @Override
        public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
            var url = context.getHttpRequest().getUrl().toString();
            if (url.startsWith(OPENAI_BASE)) {
                context.getHttpRequest().setUrl(HF_BASE + url.substring(OPENAI_BASE.length()));
            }
            return next.process();
        }
    }
}
